src/Icicle/Common/Exp/Compounds.hs

```diff
@@ -226,68 +228,75 @@ subst1 :: (Hashable n, Eq n)
 subst1 a_fresh name payload into
  = subst a_fresh (Map.singleton name payload) into
 
--- | Substitute an expression in,
--- using fresh names to avoid capture
+-- | Substitute multiple expressions.
 subst :: (Hashable n, Eq n)
       => a
-      -> Map.Map (Name n) (Exp a n p)
+      -> Map (Name n) (Exp a n p)
       -> Exp a n p
       -> Fresh n (Exp a n p)
 subst a_fresh env into
- | Map.null env
+  = let ann   = const (a_fresh, Set.empty)
+        env'  = fmap (reannotX ann) env
+        into' = reannotX ann into
+    in  reannotX fst <$> substAnn a_fresh env' into'
+
+-- | Exactly as above, but keeping track of variables
+--   under expressions as an annotation.
+substAnn :: (Hashable n, Eq n)
+         => a
+         -> Map (Name n) (Exp (Ann a n) n p)
+         -> Exp (Ann a  n) n p
+         -> Fresh n (Exp (Ann a n) n p)
+substAnn a_fresh envmap into
+ | Map.null envmap
  = return into
  | otherwise
- = subst' a_fresh env (Set.unions $ Map.elems $ Map.map freevars env) into
-
-subst'  :: (Hashable n, Eq n)
-        => a
-        -> Map.Map (Name n) (Exp a n p)
-        -> Set.Set (Name n)
-        -> Exp a n p
-        -> Fresh n (Exp a n p)
-subst' a_fresh env frees into
- = go into
+ = subst' envmap (Set.unions $ Map.elems $ Map.map freevars envmap) into
  where
-  go xx
-   = case xx of
-      XVar _ n
-       | Just payload <- Map.lookup n env
-       -> return payload
-       | otherwise
-       -> return xx
-      XApp a p q
-       -> XApp a <$> go p <*> go q
+   xVar n
+    = XVar (a_fresh, Set.singleton n) n
+   subst' env frees xx
+    = let go = subst' env frees
+      in case xx of
+       XVar _ n
+        | Just payload <- Map.lookup n env
+        -> return payload
+        | otherwise
+        -> return xx
+       XApp a p q
+        -> XApp a <$> go p <*> go q
```

Comment:
do you need to rebuild a's allvars from `go p` and `go q`, if substitution was applied in one of them?
if not, maybe explain as a comment why it works for anormal